src/cbc_sdk/platform/policy_ruleconfigs.py

```diff
@@ -272,3 +299,316 @@ def set_assignment_mode(self, mode):
         if mode not in ("REPORT", "BLOCK"):
             raise ApiError(f"invalid assignment mode: {mode}")
         self.set_parameter("WindowsAssignmentMode", mode)
+
+
+class HostBasedFirewallRuleConfig(PolicyRuleConfig):
+    """Represents a host-based firewall rule configuration in the policy."""
+    swagger_meta_file = "platform/models/policy_ruleconfig.yaml"
+
+    def __init__(self, cb, parent, model_unique_id=None, initial_data=None, force_init=False, full_doc=False):
+        """
+        Initialize the CorePreventionRuleConfig object.
+
+        Args:
+            cb (BaseAPI): Reference to API object used to communicate with the server.
+            parent (Policy): The "parent" policy of this rule configuration.
+            model_unique_id (str): ID of the rule configuration.
+            initial_data (dict): Initial data used to populate the rule configuration.
+            force_init (bool): If True, forces the object to be refreshed after constructing.  Default False.
+            full_doc (bool): If True, object is considered "fully" initialized. Default False.
+        """
+        super(HostBasedFirewallRuleConfig, self).__init__(cb, parent, model_unique_id, initial_data, force_init,
+                                                          full_doc)
+        self._rule_groups = []
+        self._rule_groups_valid = False
+
+    class FirewallRuleGroup(MutableBaseModel):
+        """Represents a group of related firewall rules."""
+        swagger_meta_file = "platform/models/firewall_rule_group.yaml"
+
+        def __init__(self, cb, initial_data):
+            """
+            Initialize the FirewallRuleGroup object.
+
+            Args:
+                cb (BaseAPI): Reference to API object used to communicate with the server.
+                initial_data (dict): Initial data used to populate the firewall rule group.
+            """
+            super(HostBasedFirewallRuleConfig.FirewallRuleGroup, self).__init__(cb, None, initial_data, False, True)
+            self._parent = None
+            self._rules = [HostBasedFirewallRuleConfig.FirewallRule(cb, d)
+                           for d in initial_data.get("rules", [])]
+
+        def _init_parent(self, parent):
+            """
+            Initialize the parent of this group and its rules.
+
+            Args:
+                parent (HostBasedFirewallRuleConfig): The parent to be set.
+            """
+            self._parent = parent
+            for rule in self._rules:
+                rule._parent = parent
+
+        def _field_updated(self, attrname):
+            """Method called whenever a field is updated."""
+            if self._parent:
+                self._parent._mark_changed()
+
+        def _flatten(self):
```

Comment:
Why do we need `_flatten` when the `_info` object contains the updated full content as python uses pass by reference? From what I can tell the location where `_flatten` is used does not modify the content only sends it to the Carbon Black Cloud backend.

Reply:
I implemented `_flatten` in both `FirewallRuleGroup` and `FirewallRule` for reasons of symmetry. It can certainly be omitted from `FirewallRule` by just tapping `_info` directly, but it's still needed in `FirewallRuleGroup`.

Reply:
`_flatten` is not needed on `FirewallRuleGroup` as it uses pass by reference and already contains all the rules same with `HostBasedFirewallRuleConfig` as this contains all the `FirewallRuleGroup` and `FirewallRule`

Reply:
The purpose of `_flatten` in `FirewallRuleGroup` is to update the _actual_ `rules` property in the new copy of `_info` with the contents of the `FirewallRule` objects. When those objects are created, they are "divorced" from the original data fed in by the server. When we want to push it _back_ to the server, we need to fill that in, or else the changes to the rules aren't going to be reflected properly. This is _especially_ the case if we add or remove a rule on the rule group.

Reply:
Maybe I missed it by I don't see use making a copy anywhere so I assumed the reference was still intact even though we are only referencing a sub component of the original json/_info

Reply:
In any event, adds and removes are _definitely_ going to require `_flatten` or something very much like it.  In neither of those cases is `_info["rules"]` actually touched, we just manipulate the `_rules` list. Since that's the case, it's easier to just do a `_flatten` whenever we need to get the changed information.  And, as part of the `save`, the new information is returned, with which we update `_info` anyway and cause the `_rules` to be regenerated whenever the rule groups list is.

Reply:
That's not necessary if we implement the adds/removes correctly as it will modify the pass by reference objects/lists

Reply:
I made this better in the code that was just committed.

Reply:
Thanks the new code looks great!